File: leetcode/problems/language-specific/go/51.n-queens.go
/*
 * @lc app=leetcode id=51 lang=golang
 *
 * [51] N-Queens
 */

// @lc code=start

// * Backtracking Solution | O(n^2) Time | O(n^2) Space

func solveNQueens(n int) [][]string {
	res := [][]string{}
	leftRow, lowerDiagonal, upperDiagonal := make([]int, n), make([]int, 2*n-1), make([]int, 2*n-1)
	board := make([]string, n)
	var emptyBoardSB strings.Builder
	for i := 0; i < n; i++ {
		emptyBoardSB.WriteString(".")
	}
	for i := range board {
		board[i] = emptyBoardSB.String()
	}

	var solveNQueensHelper func(col int)
	solveNQueensHelper = func(col int) {
		if col == n {
			res = append(res, append([]string{}, board...))
		}

		for row := 0; row < n; row++ {
			if leftRow[row] == 0 &&
				lowerDiagonal[row+col] == 0 &&
				upperDiagonal[n-1+col-row] == 0 {
				board[row] = fmt.Sprintf("%vQ%v", board[row][:col], board[row][col+1:])
				leftRow[row] = 1
				lowerDiagonal[row+col] = 1
				upperDiagonal[n-1+col-row] = 1
				solveNQueensHelper(col + 1)
				board[row] = fmt.Sprintf("%v.%v", board[row][:col], board[row][col+1:])
				leftRow[row] = 0
				lowerDiagonal[row+col] = 0
				upperDiagonal[n-1+col-row] = 0
			}
		}
	}

	solveNQueensHelper(0)
	return res
}

// @lc code=end

File: leetcode/problems/language-specific/python/51.n-queens.py
#
# @lc app=leetcode id=51 lang=python3
#
# [51] N-Queens
#

# @lc code=start

# * Backtracking Solution | O((n^2)*n!) Time | O(n^2) Space
# * n -> The given input


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        col_set = set()
        # * row + col will be constant.
        positive_diagonal_set = set()
        # * row - col will be constant.
        negative_diagonal_set = set()
        # * board -> When n = 3 -> [['.', '.', '.'], ['.', '.', '.'], ['.', '.', '.']]
        board = [["."] * n for _ in range(n)]
        res = []

        def solveNQueensHelper(row):
            if row == n:
                res.append(["".join(row_val) for row_val in board])
                return

            for col in range(n):
                if (
                    col in col_set
                    or (row + col) in positive_diagonal_set
                    or (row - col) in negative_diagonal_set
                ):
                    continue

                col_set.add(col)
                positive_diagonal_set.add(row + col)
                negative_diagonal_set.add(row - col)
                board[row][col] = "Q"
                solveNQueensHelper(row + 1)
                col_set.remove(col)
                positive_diagonal_set.remove(row + col)
                negative_diagonal_set.remove(row - col)
                board[row][col] = "."

        solveNQueensHelper(0)
        return res


# @lc code=end
